### PR TITLE
async iterator final submission

### PR DESCRIPTION
**Pull request by jlazdw**
This PR is to introduce asynchronous Iterator for getSlice And getSlices methods in FoundationDBKeyvalueStore.java ,to greatly leverage FoundationDB’s Asynchronous Iterator for on-demand data stream pulling, better memory efficiency, better parallelism for MultiQuery support. The asynchronous iterator relies on the FDB java library to take care of Completable Futures, rather than introducing additional Completable Futures in this storage plugin. 

To support this new Async Iterator mode, we introduce “iterator” mode in FoundationDBConfigOptions.java. The existing getSlice and getSlices() is in the new (default) mode called “list” mode, as the query result from FDB java library gets converted to List-based result immediately.

We also provide the fix on getSlices in the current implementation.

### CODE REVIEW

**Comment by rngcntr** [comment]
That sound like a very good improvement to me, thanks @jlazdw!
I'm kindly asking @ruweih to help me out reviewing it, because he is also working on an asynchonous implementation of `getSlice`.

**Comment by blindenvy** [comment]
+1 Thanks for the contribution @jlazdw

**Review by blindenvy — COMMENTED** [review]
+1 LGTM. Thank you so much for this contribution @jlazdw

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    @Override
+    public void close() {
+        entries.cancel();
+    }
+
+    @Override
+    public void remove() {
+        entries.remove();
+    }
+}
```
Comment (ruweih) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java:132]: Seems we can extend from existing FoundationDBRecordIterator with a new constructor accepting iterator if record instead of list of records:
https://github.com/JanusGraph/janusgraph-foundationdb/blob/e8cf024519045f501a7ad83b8090c61e9ebd82b3/src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIterator.java

In the constructor of this class, we can leverage the Guava's iterator utility to make the code cleaner:

```
            Iterator<KeyValue> entries = Iterators.filter(
                Iterators.transform(result, keyValue -> {
                    StaticBuffer key = getBuffer(db.unpack(keyValue.getKey()).getBytes(0));
                    if (selector.include(key)) {
                        return new KeyValueEntry(key, getBuffer(keyValue.getValue()));
                    } else {
                        return null;
                    }
                }),
                keyValue -> keyValue != null);
        }
       super(entries);

```

We can then reuse the base class implementation on iterator interface

Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java:132]: @ruweih:  I tried to incorporate your suggestion to the code. and I just found that the code that you suggested above can not be compiled, as  Iterator<KeyValue> entries has the type "KeyValue" and the inner transformer returns the type "KeyValueEntry". The type "KeyValue" and "KeyValueEntry" has no inheritance relationship.   

The current class "FoundationDBRecordIterator.java"  has the private member: 

private final Iterator<KeyValue> entries

It is  at the method of "next" that the "KeyValueEntry" gets constructed on the fly.  The Selector logic is applied at "FoudationDBKeyValueStore" getSlice method, before the FoundationDBRecordIterator object is constructed. https://github.com/JanusGraph/janusgraph-foundationdb/blob/master/src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java#L134. 

Therefore, if we want to keep the current FoundationDBRecordIterator's interface of using "KeyValue" and the current implementation in FoundationDBKeyValueStore's getSlice method, it is better to keep the "KeyValueEntry" construction only at the "next" method for FoundationDBRecordIteratorForAsync, just like FoundationDBRecordIterator's current implementation, instead of having the KeyValueEntry constructed at the constructor time. 

Furthermore, unlike the synchronous iterator (that is, the current FoundationDBRecordIterator.java), our asynchronous iterator will need to re-start the  FDB transaction if the transaction is too old, under "read_commit_with_write" mode, this is shown in the fetchNext() method of FoundationDBRecordIteratorAsync.java.  As a result, for a long query, we need to construct many times, the re-newed version of "entries"  with type of AsyncIterator<KeyValue>. Therefore, the  Guava's Iterator based code will have to be invoked not just in the constructor but also in the fetchNext() method, which makes the code sort of duplicated. 

If you have better way to refactor the code for the Record Iterator and the FoundationDBKeyvalueStore, please provide more details and I will incorporate your suggestion to the code.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java [hunk 2]
```diff
@@ -82,33 +112,75 @@ public synchronized void restart() {
...
+
+
+    /***
+     * Handle a potential FDBException. We simply log the exception (at the debug mode), and return false,
+     * and let's caller to: (1) if the mode is set to: serializable, then the caller will throw PermanentBackendException;
+     *     (2) if the mode is set to: either read_committed_no_write or read_committed_with_write, then the caller
+     *         will invoke re-start of the whole transaction.
+     * @return always false for now.
+     * */
+    private boolean handleFDBException(Exception e) {
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:150]: I don't quite understand the purpose of this method. When is it intended to return true, once the corresponding code has been added?
At the moment, it looks like this method should be `void` and called `logFDBException`.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java [hunk 3]
```diff
@@ -82,33 +112,75 @@ public synchronized void restart() {
...
+                }
+            }
+        }
+        return false;
     }
 
     @Override
     public synchronized void commit() throws BackendException {
         boolean failing = true;
+        int counter = 0;
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:169]: The counter is never used.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java [hunk 4]
```diff
@@ -147,78 +266,217 @@ private TransactionClose(String msg) {
     }
 
     public byte[] get(final byte[] key) throws PermanentBackendException {
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:268]: A lot of this method's complexity was removed in #45 without affecting it's semantics. It looks like you reverted most of that. Is there a real purpose behind this change or did it happen by accident?

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java [hunk 5]
```diff
@@ -147,78 +266,217 @@ private TransactionClose(String msg) {
...
-        throw new PermanentBackendException("Max transaction reset count exceeded");
+        if (failing) {
+            throw new PermanentBackendException("FDB transaction throws an exception", lastException);
+        }
+
+        return value;
     }
 
+
     public List<KeyValue> getRange(final FoundationDBRangeQuery query) throws PermanentBackendException {
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:304]: Same as for `get(...)`

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java [hunk 6]
```diff
@@ -147,78 +266,217 @@ private TransactionClose(String msg) {
...
+        final int limit = query.asKVQuery().getLimit();
+        return tx.getRange(query.getStartKeySelector(), query.getEndKeySelector(), limit,
+                false, StreamingMode.WANT_ALL).iterator();
+    }
+
+    public AsyncIterator<KeyValue> getRangeIter(FoundationDBRangeQuery query, final int skip) {
+        // Avoid using KeySelector(byte[] key, boolean orEqual, int offset) directly as stated in KeySelector.java
+        // that client code will not generally call this constructor.
+        final int limit = query.asKVQuery().getLimit();
+        KeySelector begin = query.getStartKeySelector().add(skip);
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:364]: Very nice, appreciate this idea!

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java [hunk 7]
```diff
@@ -147,78 +266,217 @@ private TransactionClose(String msg) {
...
-            for(FoundationDBRangeQuery q : retries) {
-                final KVQuery query = q.asKVQuery();
+
+        int counter = 0;
+        for (int i = 0; i < maxRuns; i++) {
+            counter++;
+
+            if (retries.size() > 0) {
+
+                List<FoundationDBRangeQuery> immutableRetries = retries.stream().collect(toList());
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:380]: Nit: The Java Stream API is rather slow compared to just creating a new list and passing `retries` as an argument to the constructor.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java [hunk 8]
```diff
@@ -147,78 +266,217 @@ private TransactionClose(String msg) {
...
+                                    }
+
+                                    // Note: the restart here will bring the code into deadlock, as restart() is a
+                                    // synchronized method and the thread to invoke this method is from a worker thread
+                                    // that serves the completable future call, which is different from the thread that
+                                    // invokes the getMultiRange call (this method) and getMultiRange is also a synchronized
+                                    // call.
+                                    //if (startTxId == txCtr.get())
+                                    //    this.restart();
+                                    resultMap.put(query, Collections.emptyList());
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:430]: Thanks for the explanation about the deadlock, sounds plausible to me!

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java [hunk 9]
```diff
@@ -117,58 +124,144 @@ public void insert(StaticBuffer key, StaticBuffer value, StoreTransaction txh, I
...
+
+            return iteratorMap;
         }
+        catch (Exception e) {
+            log.error("db={}, op=getSlicesNonAsync, tx={} throws exception", name, txh, e);
 
-        final Map<KVQuery, RecordIterator<KeyValueEntry>> iteratorMap = new HashMap<>();
-        for (Map.Entry<KVQuery, List<KeyValue>> kv : resultMap.entrySet()) {
-            iteratorMap.put(kv.getKey(), new FoundationDBRecordIterator(db, kv.getValue()));
+            if (e instanceof BackendException) {
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java:234]: See my other comment.

**Review by rngcntr — CHANGES_REQUESTED** [review]
Thanks for adding so many log outputs everywhere, that will certainly help us out hunting bugs in the future!

I've added a few comments, most of them requesting minor changes. The reason why I don't approve this PR yet is that almost none of the changes are covered in the existing tests. I think to ensure the integrity of the async methods and the iterator, we should also run the existing tests in `iterator` mode.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java (hunk 1 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java:132]: @rngcntr: thanks for the suggestion on having KeySelector to be pushed down to both synchronous iterator and the asynchronous iterator.  Let me try to make the corresponding changes and re-submit the PR.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+import com.apple.foundationdb.subspace.Subspace;
+import org.janusgraph.diskstorage.StaticBuffer;
+import org.janusgraph.diskstorage.keycolumnvalue.keyvalue.KeySelector;
+import org.janusgraph.diskstorage.keycolumnvalue.keyvalue.KeyValueEntry;
+import org.janusgraph.diskstorage.util.RecordIterator;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FoundationDBRecordIteratorForAsync implements RecordIterator<KeyValueEntry> {
```
[the remark this answers is not in the document]
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java:29]: @rngcntr: the current code does pass all of the required tests. In the PR submission, the option in FoundationDBContainer.java is set to be at:
            .set(ISOLATION_LEVEL, "read_committed_with_write")
            .set(GET_RANGE_MODE, "iterator")

Do you suggest that to modify the mvn test, so that all of the current tests run for first in the "list" mode, and then for "iterator" mode, in sequence?

[the remark this answers is not in the document]
Comment (rngcntr) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java:29]: Oh I'm sorry, the `FoundationDBContainer.java` went under my radar. But then it's the same thing just the other way around and none of the tests will use the "old" iterator now, right?
I think if both are going to be published, both should always be used in the tests. If that is better achieved in Maven or by generalizing the tests to use both configurations is up to you :smile:

[the remark this answers is not in the document]
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java]: @rngcntr: I now introduced an environment variable called "getrangemode" that can be set as "iterator" and "list" as:

-Dgetrangemode=iterator

or -Dgetrangemode=list 

to control the test execution. The environment variable is picked up by FoundationDBContainer.java and then gets pushed to FoundationDB Graph Configuration options. 

Correspondingly, I have changed the .travis.yml to augment the tests to have both "iterator" and "list" modes into the tests, effectively double the number of the tests that need to be run.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBStoreManager.java [hunk 11]
```diff
@@ -56,6 +60,9 @@
 
     private static final Logger log = LoggerFactory.getLogger(FoundationDBStoreManager.class);
 
+    public static final int ASYNC = 0, NON_ASYNC = 1;
```
[the remark this answers is not in the document]
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBStoreManager.java]: @rngcntr: I now have  in FoundationDBStoreManagers.java:

public enum RangeQueryIteratorMode { ASYNC, SYNC };

to control which get-range mode is chosen.  

I also updated README.md to explain the get-range-mode that is newly introduced in this PR.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBStoreManager.java [hunk 12]
```diff
@@ -157,29 +183,44 @@ public FoundationDBKeyValueStore openDatabase(String name) throws BackendExcepti
...
+                        log.trace("Deletion on {}: {}", mutation.getKey(), del);
+                    }
+
+                    log.debug("Total number of deletions: {}", mutationValue.getDeletions().size());
                 }
             }
         }
+        catch (Exception ex) {
+
+            if (ex instanceof BackendException) {
```
[the remark this answers is not in the document]
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBStoreManager.java]: @rngcntr: I know change the exception handling body for all of the places that handle BackendException, to be: 

try {

}
catch (BackendException e) {
     throw e;
} 
catch (Exception e) {
      throw new PermanentBackendException (e)
}

Basically, if the exception is already BackendException, we re-throw it , but if it is not an BackendException, we wrap it as a PermanentBackendException, and throw it to JanusGraphCore that handles the transaction processing.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java [hunk 13]
```diff
@@ -15,13 +21,17 @@
 import java.util.Collections;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.ArrayList;
 import java.util.Map;
+import java.util.LinkedList;
```
[the remark this answers is not in the document]
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:22]: Yes. Clean up the class import after current code change.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java (hunk 1 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java]: @rngcntr: with your suggestion above, I refactoring FoundationDBRecordIterator and FoundationDBRecordAsyncIterator, to have FoundationDBRecordAysncIterator to inherit from FoundationDBRecordIterator. The hasNext() and next() logic is in the base class of FoundationDBRecordIterator. The difference is on the fetchNext() logic.  The async iterator requires possible transaction re-start in fetchNext() logic. 

The Selector logic is implemented in the base class, FoundationDBRecordIterator. It exposes the cursor "fetched",  that allows the async iterator to set the "skip" offset when the transaction gets re-started and the getRangeIter() gets invoked again.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java (hunk 7 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java]: Yes. Simply uses:

 List<FoundationDBRangeQuery> immutableRetries = new ArrayList<> (retries);

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java (hunk 9 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java]: @rngcntr: with the latest code to refactor FoundationDBRecordIterator and FoundationDBRecordAsyncIterator, the selector logic is now pushed into the iterator itself, and thus make the FoundationDBKeyValueStore.java  getSlice () and getSlices() methods much cleaner.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java [hunk 14]
```diff
@@ -3,6 +3,12 @@
 import com.apple.foundationdb.Database;
 import com.apple.foundationdb.KeyValue;
 import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.FDBException;
+import com.apple.foundationdb.Range;
+import com.apple.foundationdb.KeyValue;
```
[the remark this answers is not in the document]
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java]: Yes. all un-used class imports get cleaned up.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java (hunk 3 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:164]: I now have the log statement to log this counter for better debugging.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java (hunk 2 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java]: Thanks for the suggestion. I changed it to be logFDBException with void return.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java (hunk 4 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:255]: First, we added the following logic to handle the exception in the get() method and getRange() method:

               if (i+1 != maxRuns) {
                    this.restart();
                } else {
                    lastException = e;
                    break;
                }

That is,  for the last iteration, we do not restart the transaction any more, as no one will pick up the restarted transaction any more 

Second, we add logFDBException so that in debug mode, we have better understanding on what FDB exception gets thrown and the nature of it:  retryable, maybe-committed, not-committed, etc.

Third, we have the finally thrown PermanentBackendException to carry the exception being thrown as:

  if (failing) {
            throw new PermanentBackendException("FDB transaction throws an exception", lastException);
}

In summary, by enriching the exception handling body, we fixed the transaction re-start logic, better logging and better exception propagation.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java (hunk 5 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:291]: See my response above on method get().

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java (hunk 1 above)
Comment (rngcntr) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java]: Sounds like a well designed solution to me! I'll most probably have another look at it later this week.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (rngcntr) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java]: Thanks!
I think it's reasonable to double the number of tests in this case.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java (hunk 4 above)
Comment (rngcntr) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:255]: What I referred to was the elimination of `boolean failing` and `byte[] value` because both are not needed and the control flow becomes much easier to read and understand if they are removed.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java (hunk 1 above)
Comment (ruweih) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java]: @jlazdw The code snippet in my previous post was from my POC of async iterator on original 0.1.0 release code, which seems having difference signature: https://github.com/JanusGraph/janusgraph-foundationdb/blob/v0.1.0/src/main/java/com/experoinc/janusgraph/diskstorage/foundationdb/FoundationDBKeyValueStore.java#L148

With the current code base, it could be further simplified to just return KeyValue:
```
            Iterator<KeyValue> entries = Iterators.filter(
                Iterators.transform(result, keyValue -> {
                    StaticBuffer key = getBuffer(db.unpack(keyValue.getKey()).getBytes(0));
                    if (selector.include(key)) {
                        return keyValue;
                    } else {
                        return null;
                    }
                }),
                keyValue -> keyValue != null);
        }
       super(entries);
```

The transaction restart logic could happen inside the inner iterators mapping method as well. I noticed there is new changes been checked in , this might not necessary as long as the common iteration logic been reused.

Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIteratorForAsync.java]: @ruweih: the difficulty that I run into to try to  use the above Iterators filter/transform that you suggested is that I need to record the cursor "fetched", so that when the transaction gets restarted in the Async iterator (inherited from the FoundationDBRecordIterator), I can skip the entries that have been already fetched. This state information will have to be recorded at the iterator level.   https://github.com/jlazdw/janusgraph-foundationdb/blob/Async_iterator_final_submission/src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIterator.java#L60

Could you help come up with the way to incorporate the state  "fetched" tracking using the filter/transform, if that is possible?

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java (hunk 4 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java]: @rngcntr: thanks for the suggestion. In the latest code change, I removed boolean failing and byte[] values, to keep the original code's flow, for both get() and getRange() two methods. Please review the code. change.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java (hunk 5 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java:274]: In the latest code change, I removed boolean failing and byte[] values, to keep the original code's flow, for both get() and getRange() two methods. Please review the code. change.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java (hunk 4 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBTx.java]: @rngcntr:  I believe that the above issue is the final issue that you have raised so far and I have addressed all of them in the latest code. If you have further comments regarding this PR, please let me know. copy @blindenvy.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIterator.java [hunk 15]
```diff
@@ -17,32 +17,52 @@
...
 
     @Override
     public KeyValueEntry next() {
-        KeyValue nextKV = entries.next();
-        StaticBuffer key = FoundationDBKeyValueStore.getBuffer(ds.unpack(nextKV.getKey()).getBytes(0));
-        StaticBuffer value = FoundationDBKeyValueStore.getBuffer(nextKV.getValue());
-        return new KeyValueEntry(key, value);
+        KeyValueEntry result = nextKeyValueEntry;
+        nextKeyValueEntry = null;
+        return result;
```
Comment (rngcntr) [inline src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIterator.java:54]: The implementation of `next()` does not match the expectation of what an Iterator does:
https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html#next--

Repeated calls of `next()` without an intermediate `hasNext()` should still continue the iteration. In other words, `next()` should not rely on `hasNext()` being called.
In addition, calling `next()` on an exhausted iteration should result in a `NoSuchElementException`, not `null`.

File: src/test/java/org/janusgraph/FoundationDBContainer.java [hunk 16]
```diff
@@ -77,13 +84,39 @@ public ModifiableConfiguration getFoundationDBConfiguration() {
...
+        ModifiableConfiguration config = buildGraphConfiguration()
             .set(STORAGE_BACKEND,"org.janusgraph.diskstorage.foundationdb.FoundationDBStoreManager")
             .set(DIRECTORY, graphName)
             .set(DROP_ON_CLEAR, false)
             .set(CLUSTER_FILE_PATH, "target/test-classes/fdb/fdb.cluster")
-            .set(ISOLATION_LEVEL, "read_committed_with_write");
+            .set(ISOLATION_LEVEL, "read_committed_with_write")
+            .set(VERSION, 620);
+
+        String getRangeMode = getAndCheckRangeModeFromTestEnvironment();
```
Comment (rngcntr) [inline src/test/java/org/janusgraph/FoundationDBContainer.java:116]: Nit: You could just keep this method as it was before and only append
```java
    .set(GET_RANGE_MODE, getAndCheckRangeModeFromTestEnvironment());
```
to the list of settings.

**Review by rngcntr — CHANGES_REQUESTED** [review]
Looks much cleaner now, thank you @jlazdw!
Only two more suggestions that came to my mind. I hope you are not annoyed yet :smile:

File: src/test/java/org/janusgraph/FoundationDBContainer.java (hunk 16 above)
Comment (jlazdw) [reply src/test/java/org/janusgraph/FoundationDBContainer.java]: Ack.

File: src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIterator.java (hunk 15 above)
Comment (jlazdw) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIterator.java]: @rngcntr: thanks for the suggestion on the iterator interface. I have made some changes following your suggestion. Over the last two days when I was making a proposal on TransactionContext to the JanusGraph developer community, https://github.com/JanusGraph/janusgraph/issues/2238,  I realize that for Threaded Transactions, the current Async Iterator will need to address multiple threads to re-set the same transaction. So I will make further changes in the next several days to handle the threaded transaction. 

I have two related questions:  (1) why in the current master branch,  the test class of FoundationDBGraphConcurrentTest gets "disabled", https://github.com/JanusGraph/janusgraph-foundationdb/blob/master/src/test/java/org/janusgraph/graphdb/foundationdb/FoundationDBGraphConcurrentTest.java#L29
(2) The related JanusGraphConcurrentTest's test cases only creates a Transaction object in each worker thread launched, it is not Threaded Transaction at all, based on the "threaded transaction" definition described at: https://docs.janusgraph.org/basics/transactions/. 

In fact, there is no test in the current JanusGraph test suites to test out threaded transactions, by searching the method invocation of "createThreadedTx()" in the entire test JanusGraph test suite. 

That brings the question on: whether the current JanusGraph really support "threaded transaction" or not?

Comment (rngcntr) [reply src/main/java/org/janusgraph/diskstorage/foundationdb/FoundationDBRecordIterator.java]: Regarding your first question: The test was already failing when we took over the project and until now, nobody started to investigate why. There's an open issue to resolve the failing test case: #28 

For the second question I get your point but I have no Idea about JanusGraph's transaction handling and I feel unable to give any qualified answer to that. If you see a serious problem here, you should maybe report it to the janusgraph-dev mailing list.

**Review by rngcntr — APPROVED** [review]

**Comment by farodin91** [comment]
@jlazdw Do you want to squash all commits? Afterwards we should be able to merge it.

**Comment by jlazdw** [comment]
@farodin91: I have squashed all the commits up to the merge point. In between my commits, there was a PR merge to master. And thus the other commit 086b8abfb1321be241f3f68a17be85a1cfcf04f1 separated my commits. Please let me know if you like me to squash further or the current squash is OK.  I am working on improving the async iterator further (including enable the FoundationDBGraphConcurrentTest related tests that have been disabled in current master). I will try to make another PR in the next two weeks.

**Comment by farodin91** [comment]
@jlazdw In the normal case, we just want to have one commit in top of the current it possible. It would be great.

**Comment by farodin91** [comment]
> I am working on improving the async iterator further (including enabling the FoundationDBGraphConcurrentTest related tests that have been disabled in current master). I will try to make another PR in the next two weeks.

Sounds awesome

**Comment by jerryjch** [comment]
@farodin91 I think we can also use the "Squash and Merge' option to merge this PR. There is no conflict to merge.

**Comment by jlazdw** [comment]
@farodin91: I talked to Jerry and he suggested above on "Squash and Merge" from your side when you merge the PR to the master. If you perform squash, could you use the following last commit message from my submission branch: https://github.com/jlazdw/janusgraph-foundationdb/commit/ec1ca28c37f019ae8c1c76aac6ba23b60be93392, which is:

"
introduce asynchronous iterator
unify the interface for synchronous iterator and asynchronous iterator
minor clean-up on code for logging and class import and refactor record iterators (sync and async)

minor changes to the code:
*re-defining enum for sync and asyn iterator
*backend exception propagation clean-up

made changes to test environment and README.md:
*introduce environment-controlled option to use iterator or list mode for testing
*add -Dgetrangemode=iterator and -Dgetrangemode=list to Travis CI jobs
*add section in README.md to explain iterator/list mode choice for GetRange related queries

"

**Review by farodin91 — APPROVED** [review]
Thank you!
